packages/gatsby/src/redux/run-sift.js

```diff
@@ -68,29 +65,128 @@ function handleFirst(siftArgs, nodes) {
   }
 }
 
-function handleMany(siftArgs, nodes, sort) {
+function handleMany(siftArgs, nodes, sort, resolvedFields) {
   let result = _.isEmpty(siftArgs)
     ? nodes
-    : sift(
-        {
+    : nodes.filter(
+        sift({
           $and: siftArgs,
-        },
-        nodes
+        })
       )
 
   if (!result || !result.length) return null
 
   // Sort results.
   if (sort) {
     // create functions that return the item to compare on
-    const sortFields = sort.fields.map(field => v => getValueAt(v, field))
+    const dottedFields = objectToDottedField(resolvedFields)
+    const dottedFieldKeys = Object.keys(dottedFields)
+    const sortFields = sort.fields
+      .map(field => {
+        if (
+          dottedFields[field] ||
+          dottedFieldKeys.some(key => field.startsWith(key))
+        ) {
+          return `__gatsby_resolved.${field}`
+        } else {
+          return field
+        }
+      })
+      .map(field => v => getValueAt(v, field))
     const sortOrder = sort.order.map(order => order.toLowerCase())
 
     result = _.orderBy(result, sortFields, sortOrder)
   }
   return result
 }
 
+// Converts a nested mongo args object into a dotted notation. acc
+// (accumulator) must be a reference to an empty object. The converted
+// fields will be added to it. E.g
+//
+// {
+//   internal: {
+//     type: {
+//       $eq: "TestNode"
+//     },
+//     content: {
+//       $regex: new MiniMatch(v)
+//     }
+//   },
+//   id: {
+//     $regex: newMiniMatch(v)
+//   }
+// }
+//
+// After execution, acc would be:
+//
+// {
+//   "internal.type": {
+//     $eq: "TestNode"
+//   },
+//   "internal.content": {
+//     $regex: new MiniMatch(v)
+//   },
+//   "id": {
+//     $regex: // as above
+//   }
+// }
+const toDottedFields = (filter, acc = {}, path = []) => {
+  Object.keys(filter).forEach(key => {
+    const value = filter[key]
+    const nextValue = _.isPlainObject(value) && value[Object.keys(value)[0]]
+    if (key === `$elemMatch`) {
+      acc[path.join(`.`)] = { [`$elemMatch`]: value }
```

Comment:
this line is different than in the `toDottedFields` loki version, is that on purpose? if not, we could also move this to `db/common`